joshua/joshua_model.py

```diff
@@ -621,7 +628,12 @@ def should_run_ensemble(tr : fdb.Transaction, ensemble_id : str) -> bool:
             # No other agents are running a test for this ensemble (is this possible?)
             return True
         if max_heartbeat_age > 10:
-            # "Kill" this seed for the presumed dead agent
+            print("Agent {} presumed dead. Attempting to steal its work.".format(
+                _get_hostname(ensemble_id, max_seed, tr)))
+
+            # If we read at snapshot isolation then an arbitrary number of agents could steal this run/seed.
+            # We only want one agent to succeed in taking over for the dead agent's run/seed.
+            tr.add_read_conflict_key(dir_ensemble_incomplete[ensemble_id]["heartbeat"][max_seed])
```

Comment:
Is it OK to rely on the snapshotted property above in line 615?
Don't we want to make sure the deletion here will always be visible?
```
props = _get_ensemble_properties(tr, ensemble_id, snapshot=True)
```

Reply:
> Is it OK to rely on the snapshotted property above in line 615?
> Don't we want to make sure the deletion here will always be visible?
> 
> ```
> props = _get_ensemble_properties(tr, ensemble_id, snapshot=True)
> ```

Never mind, I was a bit confused but @sfc-gh-anoyes explained to me offline.

Reply:
Discussed offline - snapshot isolation doesn't have to do with when mutations are visible